src/librepgp/stream-parse.c

```diff
@@ -0,0 +1,1312 @@
+    if (param) {
+        if (param->pkt.partial) {
+            param->pkt.readsrc->close(param->pkt.readsrc);
+            free(param->pkt.readsrc);
+            param->pkt.readsrc = NULL;
+        }
+
+#ifdef HAVE_BZLIB_H
+        if (param->alg == PGP_C_BZIP2) {
+            BZ2_bzDecompressEnd(&param->bz);
+        } else if ((param->alg == PGP_C_ZIP) || (param->alg == PGP_C_ZLIB))
+#endif
+        {
+            inflateEnd(&param->z);
+        }
+
+        free(src->param);
+        src->param = NULL;
+    }
+    if (src->cache) {
+        free(src->cache);
+        src->cache = NULL;
+    }
+}
+
+ssize_t
+encrypted_src_read(pgp_source_t *src, void *buf, size_t len)
+{
+    pgp_source_encrypted_param_t *param = src->param;
+    ssize_t                       read;
+    ssize_t                       mdcread;
+    ssize_t                       mdcsub;
+    bool                          parsemdc = false;
+    uint8_t                       mdcbuf[MDC_V1_SIZE];
+    uint8_t                       hash[PGP_SHA1_HASH_SIZE];
+
+    if (param == NULL) {
+        return -1;
+    }
+
+    if (src->eof) {
+        return 0;
+    }
+
+    read = src_read(param->pkt.readsrc, buf, len);
+    if (read <= 0) {
+        return read;
+    }
+
+    if (param->has_mdc) {
+        /* make sure there are always 20 bytes left on input */
+        mdcread = src_peek(param->pkt.readsrc, mdcbuf, MDC_V1_SIZE);
+        if (mdcread < MDC_V1_SIZE) {
+            if ((mdcread < 0) || (mdcread + read < MDC_V1_SIZE)) {
+                (void) fprintf(stderr, "encrypted_src_read: wrong mdc read state\n");
+                return -1;
+            }
+
+            mdcsub = MDC_V1_SIZE - mdcread;
+            memmove(&mdcbuf[mdcsub], mdcbuf, mdcread);
+            memcpy(mdcbuf, (uint8_t *) buf + read - mdcsub, mdcsub);
+            read -= mdcsub;
+            parsemdc = true;
+        }
+    }
+
+    pgp_cipher_cfb_decrypt(&param->decrypt, buf, buf, read);
+
+    if (param->has_mdc) {
+        pgp_hash_add(&param->mdc, buf, read);
+
+        if (parsemdc) {
+            pgp_cipher_cfb_decrypt(&param->decrypt, mdcbuf, mdcbuf, MDC_V1_SIZE);
+            pgp_cipher_finish(&param->decrypt);
+            pgp_hash_add(&param->mdc, mdcbuf, 2);
+            pgp_hash_finish(&param->mdc, hash);
+
+            if ((mdcbuf[0] != MDC_PKT_TAG) || (mdcbuf[1] != MDC_V1_SIZE - 2)) {
+                (void) fprintf(stderr, "encrypted_src_read: mdc header check failed\n");
+                return -1;
+            }
+
+            if (memcmp(&mdcbuf[2], hash, PGP_SHA1_HASH_SIZE) != 0) {
+                (void) fprintf(stderr, "encrypted_src_read: mdc hash check failed\n");
+                return -1;
+            }
+        }
+    }
+
+    return read;
+}
+
+void
+encrypted_src_close(pgp_source_t *src)
+{
+    pgp_source_encrypted_param_t *param = src->param;
+    if (param) {
+        FREE_ARRAY(param, symenc);
+        FREE_ARRAY(param, pubenc);
```

Comment:
I think we ought to scrub each of these as well.

Reply:
Do you mean we should change DYN_ARRAY to list, or.. ?

Reply:
No I meant that we should maybe scrub the symmetric keys with `pgp_forget`.

Reply:
Ah, got it. I don't keep symmetric keys in `pgp_sk_sesskey_t`, the idea about it is that it keeps just information from the packet itself. All keys and decryption data will go to the decrypting stream (we need just one correct session key to decrypt the stream).

However It seems I forgot to `pgp_forget` in some other places, thanks.